quinn-proto/src/connection/pacing.rs

```diff
@@ -0,0 +1,125 @@
+//! Pacing of packet transmissions.
+
+use std::time::{Duration, Instant};
+
+use tracing::{debug, warn};
+
+/// A simple token-bucket pacer. The bucket starts full and has an adjustable capacity. Once the
+/// bucket is empty, further transmission is blocked. The bucket refills at a rate slightly faster
+/// than one contestion window per RTT.
+pub struct Pacer {
+    capacity: u64,
+    tokens: u64,
+    prev: Instant,
+}
+
+impl Pacer {
+    /// Obtains a new [`Pacer`].
+    pub fn new(capacity: u64, now: Instant) -> Self {
+        Self {
+            capacity,
+            tokens: capacity,
+            prev: now,
+        }
+    }
+
+    /// Record that a packet has been transmitted.
+    pub fn on_transmit(&mut self, packet_length: u16) {
+        self.tokens = self.tokens.saturating_sub(packet_length.into())
+    }
+
+    /// Return how long we need to wait before sending a packet.
+    ///
+    /// If we can send a packet right away, this returns `None`. Otherwise, returns `Some(d)`,
+    /// where `d` is the time before this function should be called again.
+    ///
+    /// The 5/4 ratio used here comes from the suggestion that N = 1.25 in the draft IETF RFC for
+    /// QUIC.
+    pub fn delay(
+        &mut self,
+        smoothed_rtt: Duration,
+        mtu: u16,
+        window: u64,
+        now: Instant,
+    ) -> Option<Instant> {
+        debug_assert_ne!(
+            window, 0,
+            "zero-sized congestion control window is nonsense"
+        );
+
+        // if we can already send a packet, there is no need for delay
+        if self.tokens > mtu.into() {
+            return None;
+        }
+
+        // we disable pacing for extremely large windows
+        if window > u32::max_value().into() {
+            debug!("window too large, disabling pacing");
+            return None;
+        }
+
+        let time_elapsed = now.checked_duration_since(self.prev).unwrap_or_else(|| {
+            warn!("received a timestamp early than a previous recorded time, ignoring");
+            Default::default()
+        });
+
+        if smoothed_rtt.as_nanos() == 0 {
+            return None;
+        }
+
+        let elapsed_rtts = time_elapsed.as_secs_f64() / smoothed_rtt.as_secs_f64();
+        let new_tokens = window as f64 * 1.25 * elapsed_rtts;
+        self.tokens = self
+            .tokens
+            .saturating_add(new_tokens as _)
+            .min(self.capacity);
+
+        // if we can already send a packet, there is no need for delay
+        if self.tokens > mtu.into() {
+            return None;
+        }
+
+        let rtt = smoothed_rtt
+            .checked_mul(((mtu as u64).max(self.capacity) - self.tokens).min(65535) as _)
+            .unwrap_or_else(|| Duration::new(u64::max_value(), 999_999_999));
+        Some(self.prev + ((rtt / window as u32) / 5) * 4)
```

Comment:
```suggestion
        Some(self.prev + (4 * rtt / (window as u32 * 5))
```
Seems a lot clearer.

Reply:
I wrote the code the way I did to prevent any possibility of integer overflow.

Reply:
That's fair, but is sufficiently non-obvious that a comment is needed--else it's liable to inadvertently disappear in a refactoring.

Reply:
Indeed.